### DISCUSSION TITLE
InputBinding转发问题

### DISCUSSION THREAD

**Discussion by WHYBBE** [BYJRK/DotNet-Discussions#364] [WPF 开发]
@CodingOctocat 在 [discussions/357](https://github.com/BYJRK/DotNet-Discussions/discussions/357#discussioncomment-10724738) 中提到的 [IPConfig](https://github.com/CodingOctocat/IPConfig) 项目中有一个很有意思的写法

他使用一个 [InputBindingBehavior](https://github.com/CodingOctocat/IPConfig/blob/bb355dc547fe8a424d71f6a621b0788e36417096/IPConfig/Behaviors/InputBindingBehavior.cs) 类（写法也许来自于 [InputBindings work only when focused](https://stackoverflow.com/questions/23316274/inputbindings-work-only-when-focused)），将 UserControl 上的 InputBindings 转发到所属 Window 的 InputBindings 上，这样就可以忽视焦点问题直接使用快捷键 

但是会出现一个非常奇怪且不影响使用的绑定失败（“System.Windows.Data Error: Cannot find governing FrameworkElement or FrameworkContentElement for target element.”），所以有没有人能知道如何**让这个绑定失败消失** 

或者有没有更好的办法能直接在 UserControl/Page 等上使用快捷键而忽略焦点问题

**Comment by CodingOctocat** [comment]
可以解决：[WPF Error: Cannot find governing FrameworkElement for target element](https://stackoverflow.com/a/61137678/4380178)

具体方法：在 UserControl 的根元素下（如 Grid 中），添加一个代理元素

```xaml
<FrameworkElement x:Name="ProxyElement"
                  Visibility="Collapsed" />
```

然后使用 `x:Reference`：

```xaml
<UserControl.InputBindings>
    <KeyBinding Key="F5"
                Command="{Binding DataContext.RefreshCommand, Source={x:Reference ProxyElement}}"
                CommandParameter="{StaticResource True}" />
</UserControl.InputBindings>

```

对于上述链接中的高赞回答，对于此问题不起作用：https://stackoverflow.com/a/7661689/4380178。

---

如果希望更优雅的解决方案，可以使用此方法：https://stackoverflow.com/a/71256869/4380178

创建代理对象：

```cs
public class BindingProxy : Freezable
{
    #region Overrides of Freezable

    protected override Freezable CreateInstanceCore()
    {
        return new BindingProxy();
    }

    #endregion Overrides of Freezable

    // Using a DependencyProperty as the backing store for Data.  This enables animation, styling, binding, etc...
    public static readonly DependencyProperty DataProperty =
        DependencyProperty.Register(
            "Data",
            typeof(object),
            typeof(BindingProxy),
            new UIPropertyMetadata(null));

    public object Data
    {
        get => GetValue(DataProperty);
        set => SetValue(DataProperty, value);
    }
}
```

使用代理对象：

```xaml
<UserControl.Resources>
    <control:BindingProxy x:Key="Proxy"
                          Data="{Binding}" />
</UserControl.Resources>
```

```xaml
<UserControl.InputBindings>
    <KeyBinding Key="F5"
                Command="{Binding Data.RefreshCommand, Source={StaticResource Proxy}}"
                CommandParameter="{StaticResource True}" />
</UserControl.InputBindings>
```
